src/test/regress/sql/with.sql

```diff
@@ -1083,3 +1083,35 @@ WITH cte AS (
 
 RESET optimizer;
 DROP TABLE d;
+
+-- Check if sharing is disabled for a SegmentGeneral CTE to avoid deadlock if CTE is
+-- executed with 1-gang and joined with Hashed
+SET optimizer = off;
+--start_ignore
+DROP TABLE IF EXISTS d;
+DROP TABLE IF EXISTS r;
+--end_ignore
+
+CREATE TABLE d (a int, b int) DISTRIBUTED BY (a);
+INSERT INTO d VALUES (1,2),(2,3);
+CREATE TABLE r (a int, b int) DISTRIBUTED REPLICATED;
+INSERT INTO r VALUES (1,2),(3,4);
+
+EXPLAIN (COSTS off)
+WITH cte AS (
+    SELECT count(*) a FROM r
+) SELECT * FROM cte JOIN (SELECT * FROM d JOIN cte USING (a) LIMIT 1) d_join_cte USING (a);
+
+WITH cte AS (
+    SELECT count(*) a FROM r
+) SELECT * FROM cte JOIN (SELECT * FROM d JOIN cte USING (a) LIMIT 1) d_join_cte USING (a);
+
+-- Check that a correct plan is generated in case of joining two SegmentGenerals
+EXPLAIN (COSTS off)
+WITH cte AS (
+    SELECT count(*) a FROM r
+) SELECT * FROM cte JOIN (SELECT * FROM r JOIN cte USING (a) LIMIT 1) d_join_cte USING (a);
```

Comment:
do we really want to test this? it works without a patch.

Reply:
But without patch plan was differ.

Reply:
Ok, with such motivation we must add tests for all possible cases when we disable sharing of any SegmentGeneral plan.
I guess query like this is completely the same, but much simplier.
```
EXPLAIN (COSTS off)
WITH cte AS (
    SELECT a FROM r
) SELECT * FROM cte cte1 JOIN cte cte2 USING (a);
```
Again, do we really need that test?   I don't think so.

Reply:
@KnightMurloc, @trexxet2  Maybe we omit that test?

Reply:
I think it makes sense to remove it since many plans have changed with CTE with the SegmentGeneral locus type.